Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,136 @@
+//
+// This source file is part of Prelude, an open source project by Wayfair
+//
+// Copyright (c) 2018 Wayfair, LLC.
+// Licensed under the 2-Clause BSD License
+//
+// See LICENSE.md for license information
+//
+
+/// protocol that represents a rule that can be used to “refine” values of type `RefinedType`. Implement the `static Refinement.isValid(_:)` function to describe which values are allowed.
+public protocol Refinement {
+    /// the type being “refined”
+    associatedtype RefinedType
+
+    /// implement this function to describe which values of type `RefinedType` should be allowed
+    ///
+    /// - Parameter value: a value of type `RefinedType`
+    /// - Returns: `true` if this rule should allow `value`
+    static func isValid(_ value: RefinedType) -> Bool
+}
+
+/// error that is thrown when a value fails a refinement. Examine `localizedDescription` to determine the details of the failure
+public struct RefinementError: Error {
+    public let localizedDescription: String
+
+    init(_ localizedDescription: String) {
+        self.localizedDescription = localizedDescription
+    }
+}
+
+/// a container for a “refined” type. The type parameter `A` represents the underlying value (eg. `Int`). The type parameter `R` represents the “refinement” (eg. `NonZero`, `NotNegative`, `GreaterThanFifty`, etc.)
+public struct Refined<A, R: Refinement> where A == R.RefinedType {
+    /// the underlying refined value. This value is guaranteed to satisfy the refinement represented by `R`
+    public let value: A
+
+    /// initialize a `Refined` value with the proposed underlying value `value`. If the value satisfies the rule  represented by `R`, then this initializer will succeed. Otherwise, this initializer will throw a `RefinementError` describing the failure
+    ///
+    /// - Parameter value: the proposed underlying value
+    /// - Throws: a `RefinementError` describing any failure
+    public init(_ value: A) throws {
+        guard R.isValid(value) else {
+            throw RefinementError("the value “\(value)” doesn’t satisfy the refinement “\(R.self)”")
+        }
+        self.value = value
+    }
+}
+
+extension Refined: Equatable where A: Equatable { }
+
+extension Refined: Hashable where A: Hashable { }
+
+public extension Refinement {
+    /// convenient way initialize a value of type `Refined<RefinedType, Self>` given a proposed value. This is equivalent to calling `try? Refined<RefinedType, Self>.init(value)`, but much easier to type
+    ///
+    /// - Parameter value: the proposed value
+    /// - Returns: a value of type `Refined<RefinedType, Self>`, or nil
+    static func of(_ value: RefinedType) -> Refined<RefinedType, Self>? {
+        return try? .init(value)
+    }
```

Comment:
Do you think it's valuable to add some conformances to ExpressibleByStringLiteral/IntegerLiteral/etc where A also conforms to those protocols? We wouldn't be able to throw or return an optional thing here if we did, so it might not be safe at runtime to allow people to do `let x: NonZeroDouble = 0` since it'd have to fatalError, but the convenience might be worth the tradeoff. What do you think?

Reply:
It’s tempting but I think I come down on the “don’t add them” side myself. In the long run having a line of code like `let foo: NonZeroDouble = 0` possibly result in a `fatalError` is just too weird. I don’t think I want to live in that world! 😆

Reply:
Fair enough :)